File: app/src/main/java/ru/geekbrains/notes/observe/Publisher.java
package ru.geekbrains.notes.observe;

import java.util.ArrayList;
import java.util.List;

import ru.geekbrains.notes.card.CardData;

public class Publisher {
    private List<Observer> observers;

    public Publisher() {
        this.observers = new ArrayList<Observer>();
    }

    public void subscribe(Observer observer) {
        observers.add(observer);
    }

    public void unsubscribe(Observer observer) {
        observers.remove(observer);
    }

    public void notifyTask(CardData cardData) {
        for (Observer observer : observers) {
            observer.updateState(cardData);
        }
    }
}

File: app/src/main/java/ru/geekbrains/notes/presenter/observation/Publisher.kt
package ru.geekbrains.notes.presenter.observation

import ru.geekbrains.notes.model.CardData

object Publisher {
    private val observers = ArrayList<Observer>()
    fun subscribe(observer: Observer) {
        observers.add(observer)
    }

    private fun unsubscribe(observer: Observer) {
        observers.remove(observer)
    }

    fun notifyTask(cardData: CardData) {
        for (observer in observers) {
            observer.updateState(cardData)
            unsubscribe(observer)
        }
    }
}
